### PR TITLE
api: remove hardcoded type urls part.1

### LINKED ISSUE #10209: hardcoded v2 type urls in xds requests
*Title*: *hardcoded v2 type urls in xds requests*

*Description*:
In order to migrate to v3 API, `loadTypeUrl` functions are used to determine type url in discovery requests. However, I found that some files are still using v2 type url constants in [this file](https://github.com/envoyproxy/envoy/blob/master/source/common/config/resources.h). This means that even envoy is configured to use v3, it will still send requests for v2 resources.

Correct me if there is any magic actually doing the conversion somewhere.

[optional *Relevant Links*:]
https://github.com/envoyproxy/envoy/blob/334724c8cbbdffcafcfb0590a9dd1d9bd71917e0/source/common/router/scoped_rds.cc#L243
https://github.com/envoyproxy/envoy/blob/334724c8cbbdffcafcfb0590a9dd1d9bd71917e0/source/common/router/scoped_rds.cc#L258
https://github.com/envoyproxy/envoy/blob/334724c8cbbdffcafcfb0590a9dd1d9bd71917e0/source/common/upstream/cds_api_impl.cc#L68-L70
https://github.com/envoyproxy/envoy/blob/334724c8cbbdffcafcfb0590a9dd1d9bd71917e0/source/common/upstream/cluster_manager_impl.cc#L150-L155
https://github.com/envoyproxy/envoy/blob/334724c8cbbdffcafcfb0590a9dd1d9bd71917e0/source/common/upstream/cluster_manager_impl.cc#L752-L754
https://github.com/envoyproxy/envoy/blob/334724c8cbbdffcafcfb0590a9dd1d9bd71917e0/source/server/lds_api.cc#L43-L45
https://github.com/envoyproxy/envoy/blob/334724c8cbbdffcafcfb0590a9dd1d9bd71917e0/source/server/server.cc#L526-L535

### PR DESCRIPTION
**Pull request by Shikugawa**
Signed-off-by: shikugawa <rei@tetrate.io>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description: #10209. There are hardcoded type URLs in resources.h. In this PR, I removed all of the dependencies of those. Now, codes that depend on that are only test code. In the future, I will remove that from test codes and completely destroy that.

Risk Level: Low
Testing: N/A
Docs Changes: N/A
Release Notes: N/A
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

File: source/common/upstream/cluster_manager_impl.cc [hunk 1]
```diff
@@ -143,16 +142,31 @@ void ClusterManagerInitHelper::maybeFinishInitialize() {
...
-      if (cm_.adsMux() == nullptr ||
-          cm_.adsMux()->paused(Config::TypeUrl::get().ClusterLoadAssignment)) {
+      if (cm_.adsMux() == nullptr) {
         initializeSecondaryClusters();
+        return;
       } else {
-        cm_.adsMux()->pause(Config::TypeUrl::get().ClusterLoadAssignment);
-        Cleanup eds_resume(
-            [this] { cm_.adsMux()->resume(Config::TypeUrl::get().ClusterLoadAssignment); });
+        bool cluster_load_assignment_paused = false;
```
Comment (htuch) [inline source/common/upstream/cluster_manager_impl.cc]: I'm not a fan of conflating this behavioral change with the rest of the PR; this is potentially dangerous and might break things, the rest is just simple movement. I would prefer two PRs, thanks.

File: source/server/lds_api.cc [hunk 2]
```diff
@@ -41,9 +43,16 @@ void LdsApiImpl::onConfigUpdate(
     const std::string& system_version_info) {
   std::unique_ptr<Cleanup> maybe_eds_resume;
   if (cm_.adsMux()) {
-    cm_.adsMux()->pause(Config::TypeUrl::get().RouteConfiguration);
-    maybe_eds_resume = std::make_unique<Cleanup>(
-        [this] { cm_.adsMux()->resume(Config::TypeUrl::get().RouteConfiguration); });
+    const auto target_type_urls =
```
Comment (htuch) [inline source/server/lds_api.cc]: Same, etc.

File: source/common/upstream/eds.cc [hunk 3]
```diff
@@ -22,6 +22,8 @@ EdsClusterImpl::EdsClusterImpl(
     Stats::ScopePtr&& stats_scope, bool added_via_api)
     : BaseDynamicClusterImpl(cluster, runtime, factory_context, std::move(stats_scope),
                              added_via_api),
+      Envoy::Config::SubscriptionBase<envoy::config::endpoint::v3::ClusterLoadAssignment>(
```
Comment (htuch) [inline source/common/upstream/eds.cc]: FWIW, we're now slowly drifting into the space of true multiple inheritance, which is used somewhat sparingly in Envoy. Not sure if the horse has fully bolted here, @mattklein123?

Comment (mattklein123) [reply source/common/upstream/eds.cc]: TBH I don't really have a strong opinion on this. I guess I'm of an opinion that our default should be no, but if there is a good case for it I don't see any major issue. cc @envoyproxy/maintainers

**Comment by Shikugawa** [comment]
@htuch That is, We should divide this PR into two, shouldn't we? First, basic cleanup of `SubscriptionBase`,  `getResourceName` and so on. Second, v2 and v3 resource pause and resume with some tests. OK?

**Comment by htuch** [comment]
@Shikugawa yep.

**Comment by Shikugawa** [comment]
@htuch Destroy hardcoded type URLs on this PR. I will create Part.2 that includes hardcoded type URLs removal and some tests with v3 resource names.

**Comment by mattklein123** [comment]
@lizan can you take a look at clang-tidy here? I think this PR should have the fix for the headers issue so I think there might be a different issue?

/wait

**Comment by Shikugawa** [comment]
@mattklein123 @htuch LGTM?

**Comment by marcomagdy** [comment]
@Shikugawa @htuch 
How does the template
```cpp
template <typename Current> struct SubscriptionBase
```
Differ from the one in `discovery_service_base.h`?

Both have the same namespace and looks like they do the same thing more or less??
I'm asking because this is showing up in clang-tidy and failing CI for other PRs.

**Comment by lizan** [comment]
@marcomagdy This PR removees `discovery_service_base.h`, so if you see this failing CI for other PRs, they likely need merge master. Azure pipelines runs on merge commit so it could cause other PRs failing.

**Comment by marcomagdy** [comment]
Interestingly, my PR failed CI _after_ I merged from master.

**Pull request merged**